File: src/python/graph/3_simplify_passive_components.py
# file: 3_simplify_passive_components.py
# Simplify graph: replace passive component pins with component node

import networkx as nx # read_gml, write_gml
import re            # match
import sys           # argv

def is_passive(node):
    return re.match(r'^[RCL]\d+', node) is not None

def main():
    input_file, output_file = sys.argv[1], sys.argv[2]
    G = nx.read_gml(input_file)
    H = G.copy()
    to_simplify = {}
    for node in G.nodes():
        if is_passive(node):
            pins = [nbr for nbr in G.neighbors(node) if nbr.startswith(f"{node}_")]
            if len(pins) == 2: to_simplify[node] = pins
    for comp, pins in to_simplify.items():
        n1s, n2s = [n for n in G.neighbors(pins[0]) if n != comp], [n for n in G.neighbors(pins[1]) if n != comp]
        if n1s and n2s:
            for n in n1s: H.add_edge(comp, n)
            for n in n2s: H.add_edge(comp, n)
            H.remove_node(pins[0]); H.remove_node(pins[1])
    nx.write_gml(H, output_file)

if __name__ == "__main__":
    main()

File: src/typescript/graph/3_simplify_passive_components.ts
import fs from 'fs';
import Graph from 'graphology';

export function simplifyPassiveComponents(inputPath: string, outputPath: string): Graph {
    console.log(`[GRAPH] 3. Simplifying passive components (R, C, L) in ${path.basename(inputPath)}`);
    const G = new Graph();
    if (!fs.existsSync(inputPath)) return G;
    G.import(JSON.parse(fs.readFileSync(inputPath, 'utf8')));
    
    const H = G.copy();
    const passives: string[] = [];
    H.forEachNode(node => {
        if (/^[RCL]\d+/.test(node)) {
            const pins = H.neighbors(node).filter(nbr => nbr.startsWith(`${node}_`));
            if (pins.length === 2) {
                passives.push(node);
            }
        }
    });

    console.log(`[GRAPH] Found ${passives.length} candidate passive components.`);
    let simplifiedCount = 0;
    for (const comp of passives) {
        const pins = H.neighbors(comp).filter(nbr => nbr.startsWith(`${comp}_`));
        const p1 = pins[0];
        const p2 = pins[1];
        const neighbors1 = H.neighbors(p1).filter(n => n !== comp);
        const neighbors2 = H.neighbors(p2).filter(n => n !== comp);

        if (neighbors1.length > 0 && neighbors2.length > 0) {
            neighbors1.forEach(n => H.addEdge(comp, n));
            neighbors2.forEach(n => H.addEdge(comp, n));
            H.dropNode(p1);
            H.dropNode(p2);
            simplifiedCount++;
        }
    }
    console.log(`[GRAPH] Simplified ${simplifiedCount} passive components.`);
    fs.writeFileSync(outputPath, JSON.stringify(H.export()));
    return H;
}
